### ISSUE TITLE
CSV-Import doesn't lookup records by "name" field

### ISSUE THREAD

**Issue by MrGeneration** [zammad/zammad#2658]
### Infos:

* Used Zammad version: 3.0
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any
* Ticket-ID: #1048469

### Expected behavior:

When importing CSV-Records with doubled contents like (`Zammad`, `ZAMMAD`), Zammad will throw an error (or simply skip the entry and tell the user) while importing the rest of the entries.

### Actual behavior:

When importing CSV-Records with doubled contents like (`Zammad`, `ZAMMAD`), Zammad will throw an exception stating that the transaction is still open. Zammad is not able to insert the further entries and will throw an error, that's not directly attached with this (because the statement is still open).

Note: The statement is closed after the import has "finished" (so no big harm here).

Example:
```ruby
ActiveRecord::StatementInvalid (PG::InFailedSqlTransaction: ERROR: current transaction is aborted, commands ignored until end of transaction block)
: SELECT "settings".* FROM "settings" WHERE "settings"."area" = $1 ORDER BY "settings"."name" ASC
```

### Steps to reproduce the behavior:

```csv
name,shared,domain,domain_assignment,active,note
Zammad,false,,false,true,
ZAMMAD,false,,false,true,
```

* Use the above csv-content to import an organization
* Zammad will fail

### Scope:

This also affects Issue #2659 .

As far as we're aware, this will affect `Organizations` and `TextModules`.
This should be fixed by the following (Thorsten and I checked before hand if `name` is used in other objects where we could shoot in our feet):

```diff
diff --git a/app/models/concerns/can_csv_import.rb b/app/models/concerns/can_csv_import.rb
index f8b17e6d0..e54e2f6d2 100644
--- a/app/models/concerns/can_csv_import.rb
+++ b/app/models/concerns/can_csv_import.rb
@@ -187,7 +187,7 @@ returns
           next if attributes[lookup_by].blank?
 
           params = {}
-          params[lookup_by] = if %i[email login].include?(lookup_by)
+          params[lookup_by] = if %i[name email login].include?(lookup_by)
                                 attributes[lookup_by].downcase
                               else
                                 attributes[lookup_by]
```

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by martini** [comment]
After a discussion with @mrgeneration we agreed to raise an error (not skip singel rows) if the import contains 2 or more duplicate rows to let he user know that there are issues with the complete import.

**Comment by rlue** [comment]
I have a fix for this, and it's almost ready for QA—just want to address #2659 at the same time.

A couple notes:

1. > ```diff
   > -          params[lookup_by] = if %i[email login].include?(lookup_by)
   > +          params[lookup_by] = if %i[name email login].include?(lookup_by)
   >                                  attributes[lookup_by].downcase
   > ```

   This doesn't actually work with the sample CSV file provided above. The first entry creates an organization with the name 'Zammad', then when we reach the second record, it tries `Organization.lookup(name: 'ZAMMAD'.downcase)` and finds nothing. I solved this by using an extra `if` branch to handle case-insensitive searches:

   ```ruby
   record = if lookup_by.in?(%i[name])
              where("LOWER(#{lookup_by}) = ?", attributes[lookup_by].downcase).first  # rubocop:disable Rails/FindBy
            elsif lookup_by.in?(%i[email login])
              lookup(attributes.slice(lookup_by).transform_values!(&:downcase))
            else
              lookup(attributes.slice(lookup_by))
            end
   ```

   `.where.first` has poorer performance (doesn't use caching), and isn't necessary for `email` and `login`, since those are downcased before being stored to the database.

2. > **Expected behavior:** When importing CSV-Records with doubled contents like (Zammad, ZAMMAD), Zammad will throw an error (or simply skip the entry and tell the user) **while importing the rest of the entries.**

   I asked @martini about this, and his opinion was that the entire import should fail, "Because it's a bad import file. And usally more stuff will be wrong in the file. So in order to prevent the user to do bad stuff... we can block the import completely."

   I prepared the MR according to Martin's instructions, but if either @thorsteneckel or @MrGeneration disagrees, I'm happy to take a second look.

   **EDIT:** Whoops, missed Martin's comment above. -\_-'
